File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int profit = 0, min_so_far = Integer.MAX_VALUE;
        
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < min_so_far) {
                min_so_far = prices[i];
            } else {
                profit = Math.max(profit, prices[i] - min_so_far);
            }
        }
        
        return profit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_so_far = 10004
        profit = 0

        for price in prices:
            if price < min_so_far:
                min_so_far = price

            profit = max(profit, price - min_so_far)

        return profit